libnmstate/nm/port_vlan.py

```diff
@@ -0,0 +1,238 @@
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Lesser General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public License
+# along with this program. If not, see <https://www.gnu.org/licenses/>.
+#
+
+
+from libnmstate.nm import nmclient
+from libnmstate.schema import LinuxBridge as LB
+
+
+class PortVlanFilter(object):
+    pvid = 'pvid'
+    min_range = 'vlan-range-min'
+    max_range = 'vlan-range-max'
+    untagged = 'untagged'
+
+    def __init__(self, trunk_tags, tag, is_native_vlan=False):
+        self._trunk_ports = trunk_tags
+        self._tag = tag
+        self._is_native_vlan = is_native_vlan
+        self._port_type = (
+            LB.Port.Vlan.TRUNK_TYPE if trunk_tags else LB.Port.Vlan.ACCESS_TYPE
+        )
+
+    @property
+    def trunk_ports(self):
+        return self._trunk_ports
+
+    @property
+    def tag(self):
+        return self._tag
+
+    @property
+    def enable_native_vlan(self):
+        return self._is_native_vlan
+
+    @property
+    def port_type(self):
+        return self._port_type
+
+    def generate(self):
+        """
+        Generate a list of NM.BridgeVlan objects from the encapsulated
+        PortVlanFilter data
+        """
+
+        port_vlan_config = []
+        if self.port_type == LB.Port.Vlan.TRUNK_TYPE:
+            for trunk_port in self.trunk_ports:
+                port_vlan_config.append(
+                    self._generate_vlan_trunk_port_config(trunk_port)
+                )
+            if self.enable_native_vlan and self.tag:
+                port_vlan_config.append(
+                    self._generate_vlan_access_port_config(self.tag)
+                )
+        elif self.port_type == LB.Port.Vlan.ACCESS_TYPE and self.tag:
+            port_vlan_config.append(
+                self._generate_vlan_access_port_config(self.tag)
+            )
+
+        return port_vlan_config
+
+    def to_dict(self):
+        """
+        Get the port vlan filtering configuration in dict format - e.g. in yaml
+        format:
+
+        - name: eth1
+          vlan:
+            type: trunk
+            trunk-tags:
+              - id: 101
+              - id-range:
+                  min: 200
+                  max: 4095
+            tag: 100
+            enable-native: true
+        """
+
+        port_vlan_state = {
+            LB.Port.Vlan.TYPE: self.port_type,
+            LB.Port.Vlan.TRUNK_TAGS: self.trunk_ports,
+        }
+        if self.tag:
+            port_vlan_state[LB.Port.Vlan.TAG] = self.tag
+        if self.port_type == LB.Port.Vlan.TRUNK_TYPE:
+            port_vlan_state[
+                LB.Port.Vlan.ENABLE_NATIVE
+            ] = self.enable_native_vlan
+        return port_vlan_state
+
+    @staticmethod
```

Comment:
Factory methods should be class methods (`@classmethod`).

Reply:
Done.

Reply:
No plans to inherit from PortVlanFilter - for me static makes more sense - that's how it was originally. 

@tyll : are you OK with having this as static ? 

I can go either way.

Reply:
Stopped using it w/ static / class methods, hence, this no longer applies.